### PR TITLE
Generate rule for hs-boot files

### PR DESCRIPTION
**Pull request by GuillaumeGen**
This addresses #16

### CODE REVIEW

File: gazelle_haskell_modules/rule_generation.go [hunk 1]
```diff
@@ -398,17 +398,29 @@ type ModuleImport struct {
...
+	aux := tempMod.ModuleName
 	pkgName := ""
 	if len(aux) > 1 {
 		pkgName = aux[0]
 	}
+	suffix := ""
+	// We add the suffix `.hs-boot` to the module name,
+	// when this module comes from an `hs-boot` file.
+	if tempMod.IsSourceImported {
+		suffix = ".hs-boot"
```
Comment (googleson78) [inline gazelle_haskell_modules/rule_generation.go:420]: I think you should factor out `".hs-boot"` in a similar manner to the other "global constants", e.g. 
```golang
const PRIVATE_ATTR_MODULE_LABELS = "module_labels"
```

File: gazelle_haskell_modules/dependency_resolution.go [hunk 2]
```diff
@@ -135,27 +136,44 @@ func findModuleLabelByModuleName(
 	libs []label.Label,
 ) (*label.Label, error) {
 	spec := moduleByModuleImportSpec(moduleImport)
+	isBootDep := false
+	// If the module we are considering comes from an `hs-boot` file,
+	// we store this in the `isBootDep` variable and then
+	// behave as if we were considering the standard version of ther module.
+	if strings.HasSuffix(spec.Imp, ".hs-boot") {
```
Comment (googleson78) [inline gazelle_haskell_modules/dependency_resolution.go:143]: The `strings.HasSuffix` check is used multiple times, it might be worth factoring out?

File: himportscan/src/HImportScan/ImportScanner.hs [hunk 3]
```diff
@@ -55,9 +60,15 @@ instance Aeson.ToJSON ScannedImports where
...
+-- the type `ImportMethod`, isomorphic to `Bool`
+-- is casted during the translation to JSON.
 instance Aeson.ToJSON ModuleImport where
-  toJSON (ModuleImport maybePackageName moduleName) =
-    Aeson.toJSON $ catMaybes [maybePackageName, Just moduleName]
+  toJSON (ModuleImport importMethod maybePackageName moduleName) =
+    Aeson.object
+      [ ("isSourceImported", Aeson.Bool (importMethod == SourceImport))
+      , ("moduleName", Aeson.toJSON $ catMaybes [maybePackageName, Just moduleName])
+      ]
```
Comment (googleson78) [inline himportscan/src/HImportScan/ImportScanner.hs:71]: Since you're already using an object here anyways, I think it would be an overall improvement to do the same for `moduleName` as well and make this a flat object with three keys. It also simplifies `UnmarshalJSON`, I think.

File: himportscan/tests/HImportScan/ImportScannerSpec.hs [hunk 4]
```diff
@@ -10,7 +10,7 @@ import qualified Data.Set as Set
 import Data.String.QQ (s)
```
Comment (googleson78) [inline himportscan/tests/HImportScan/ImportScannerSpec.hs:10]: Might be nice to add a test with a `{-# SOURCE #-}` import here.

File: gazelle_haskell_modules/rule_generation.go [hunk 5]
```diff
@@ -398,17 +398,27 @@ type ModuleImport struct {
...
 	if len(aux) > 1 {
 		pkgName = aux[0]
 	}
+	suffix := ""
+	if tempMod.IsSourceImported {
+		suffix = "-boot"
+	}
 	moduleImport.PackageName = pkgName
-	moduleImport.ModuleName = aux[len(aux)-1]
+	moduleImport.ModuleName = aux[len(aux)-1] + suffix
```
Comment (facundominguez) [inline gazelle_haskell_modules/rule_generation.go]: Why not augment `ModuleImport` with a boolean `IsSourceImported` field instead of putting the information in unstructured fashion as a string suffix?

File: gazelle_haskell_modules/rule_generation.go [hunk 6]
```diff
@@ -398,17 +398,29 @@ type ModuleImport struct {
...
 	}
+	suffix := ""
+	// We add the suffix `.hs-boot` to the module name,
+	// when this module comes from an `hs-boot` file.
+	if tempMod.IsSourceImported {
+		suffix = ".hs-boot"
+	}
 	moduleImport.PackageName = pkgName
-	moduleImport.ModuleName = aux[len(aux)-1]
+	moduleImport.ModuleName = aux[len(aux)-1] + suffix
```
Comment (googleson78) [inline gazelle_haskell_modules/rule_generation.go:423]: Did you consider permanently storing the `IsSourceImported` field in `ModuleImport` instead of adding it to the string and using that as the future source of information? It might be easier to understand/not forget in the long run, albeit possibly more of a hassle to implement.

File: gazelle_haskell_modules/dependency_resolution.go [hunk 7]
```diff
@@ -135,27 +136,37 @@ func findModuleLabelByModuleName(
...
-					"Multiple rules define %q:%s in %v: %v and %v",
-					moduleImport.PackageName,
-					moduleImport.ModuleName,
-					intersection,
-					foundLabel,
-					r.Label,
-				)
-			} else {
-				foundLabel = r.Label
+		if !strings.HasSuffix(r.Label.Name, "-boot") {
```
Comment (facundominguez) [inline gazelle_haskell_modules/dependency_resolution.go]: There is some new behavior added to this function that isn't being documented.

It looks like if imports are source imports, the label of a boot module should be returned, but this is skipping all boot modules.

File: himportscan/src/HImportScan/ImportScanner.hs [hunk 8]
```diff
@@ -55,9 +59,15 @@ instance Aeson.ToJSON ScannedImports where
...
+-- Since it is easier to store in JSON format,
+-- the type `ImportMethod`, isomorphic to `Bool`
+-- is casted during the translation to JSON.
 instance Aeson.ToJSON ModuleImport where
-  toJSON (ModuleImport maybePackageName moduleName) =
-    Aeson.toJSON $ catMaybes [maybePackageName, Just moduleName]
+  toJSON (ModuleImport importMethod maybePackageName moduleName) =
+    Aeson.object
+      [ ("isSourceImported", Aeson.Bool (importMethod == SourceImport))
+      , ("moduleName", Aeson.toJSON $ catMaybes [maybePackageName, Just moduleName])
```
Comment (facundominguez) [inline himportscan/src/HImportScan/ImportScanner.hs:70]: Maybe we can make `isSourceImported` present only when the module is source imported. Otherwise the encoding grows for all imports unnecessarily. Booleans are initialized to false in go by default: https://go.dev/tour/basics/12

We could do something similar for the package name, but that can be in a separate PR if you want.

File: tests/alternative-deps/package-a/BUILD.bazel [hunk 9]
```diff
@@ -74,3 +71,65 @@ sh_test(
     name = "test-another-haskell-binary",
     srcs = [":another-haskell-binary"],
 )
+
+# rule generated by gazelle_haskell_modules
+haskell_module(
```
Comment (facundominguez) [inline tests/alternative-deps/package-a/BUILD.bazel]: Are these generated rules included here by accident?

**Review by facundominguez — COMMENTED** [review]
Thanks @GuillaumeGen!
I think that the code could rely a bit less on checking suffixes if `ModuleImport` had an extra boolean field. And perhaps it would all blend better if when indexing module imports we include in the key the fact that we are indexing a source import too.

File: gazelle_haskell_modules/dependency_resolution.go [hunk 10]
```diff
@@ -156,6 +156,7 @@ func findModuleLabelByModuleName(
 		}
 	}
 	if foundLabel.Name != "" {
+		foundLabel.Name = foundLabel.Name
```
Comment (facundominguez) [inline gazelle_haskell_modules/dependency_resolution.go:159]:
```suggestion
```

**Review by facundominguez — APPROVED** [review]
Looks solid to me! Thanks @GuillaumeGen!
